share/smack/include/pthread.h

```diff
@@ -0,0 +1,71 @@
+//
```

Comment:
I still do not understand where the fundamental conflict arises. Here ares some concrete questions; please be mindful of the perhaps subtle distinction I am trying to make between the pthread API *headers* and *implementations* — including the possibility that the line between the two is blurred:

1. Do the *headers* of the pthreads API determine the layout of the `pthread_t` type? Or is that determined by pthreads API *implementations*? Or are the *headers* dependent on the *implementation*? In other words, is the layout of `pthread_t` type necessarily fixed when modules are compiled against the pthread API headers? Are there no “generic” pthread API headers to compile against? (On asking this, I realize that the pthread API headers themselves must be implementation dependent, but I want to be sure we are clear on these points.)

2. Now, even supposing we are forced to determine the layout of pthread API types, e.g., `pthread_t`, at compile time, does anything prevent us from linking a module which implements the pthread API with possibly conflicting pthread API types?

For example, suppose the pthread API headers we compile our program against on a given system say that `pthread_t` is 20 bytes long. Then suppose Monty’s implementation — using custom pthread API headers — says that `pthread_t` is 4 bytes long. Let’s suppose for the moment that Monty’s `pthread_t` is smaller than or equal to in size than the program’s. If we were then to link Monty’s implementation with that program, what goes wrong? Perhaps we don’t want to execute it, but from the standpoint of our Boogie encoding, is there any problem with it?

I have tried such type-incompatible linking out just to see what happens, and so far I don’t see the obstacle. Consider the following code:
````
// a.h
struct S {
  int a1;
  int a2;
};

// b.h
struct S {
  int b1;
};

// a.c
#include "a.h"
#include "stdio.h"
void foo(struct S);
int main(void) {
  struct S s;
  foo(s);
  printf("World!\n");
}

// b.c
#include "b.h"
#include "stdio.h"
void foo(struct S s) {
  printf("Hello!\n");
  s.b1 = 1;
}
````
This all happily compiles, links, and executes.
````
$ clang -c -emit-llvm a.c
$ clang -c -emit-llvm b.c
$ llvm-link a.bc b.bc -o c.bc
$ clang c.bc 
$ ./a.out 
Hello!
World!
````

So why are we insisting on fixing the pthread API headers that the program compiles against?

(I am not necessarily suggesting, yet, that there is not reason. It is just that I do not yet see the reason.)